lib/cn/cn_tree_cursor.c

```diff
@@ -0,0 +1,769 @@
+        if (ev(!k))
+            return merr(ENOMEM);
+
+        kvset_get_ref(kvset);
+        k->kvset = kvset;
+
+        lcur->cnlc_dgen_lo = dgen;
+    }
+
+    return 0;
+}
+
+static void
+kvref_tab_putref(void *arg)
+{
+    struct kvref *k = arg;
+
+    if (k->kvset) {
+        kvset_put_ref(k->kvset);
+        k->kvset = 0;
+    }
+}
+
+MTF_STATIC merr_t
+cn_tree_lcur_init(
+    struct cn_level_cursor *lcur)
+{
+    struct cn_cursor       *cncur = lcur->cnlc_cncur;
+    struct table           *tab = lcur->cnlc_kvref_tab;
+    struct element_source **esrc;
+
+    struct workqueue_struct *maint_wq = cn_get_maint_wq(cncur->cncur_cn);
+
+    uint iterc = table_len(tab);
+    uint i, bh_align = 16;
+    size_t bh_max_cnt;
+    merr_t err;
+
+    lcur->cnlc_iterc = 0;
+
+    /* Grow element source array if necessary.
+     */
+    if (iterc > lcur->cnlc_esrcc) {
+        size_t cnt = ALIGN(iterc, 32);
+        void *p = realloc(lcur->cnlc_esrcv, cnt * sizeof(*lcur->cnlc_esrcv));
+
+        if (ev(!p))
+            return merr(ENOMEM);
+
+        lcur->cnlc_esrcc = cnt;
+        lcur->cnlc_esrcv = p;
+    }
+
+    esrc = lcur->cnlc_esrcv;
+
+    for (i = 0; i < iterc; i++) {
+        struct kvref *k = table_at(tab, i);
+        struct kv_iterator *it;
+
+        err = kvset_iter_create(k->kvset, NULL, maint_wq, NULL, cncur->cncur_flags, &it);
+        if (ev(err))
+            return err;
+
+        *esrc++ = kvset_iter_es_get(it);
+        ++lcur->cnlc_iterc;
+    }
+
+    if (!lcur->cnlc_iterc)
+        return 0;
+
+    /* Grow binheap if necessary
+     */
+    bh_max_cnt = ALIGN(lcur->cnlc_iterc, bh_align);
+    if (bh_max_cnt > lcur->cnlc_bh_max_cnt) {
+        bin_heap2_destroy(lcur->cnlc_bh);
+        lcur->cnlc_bh = 0;
+
+        lcur->cnlc_bh_max_cnt = bh_max_cnt;
+        err = bin_heap2_create(lcur->cnlc_bh_max_cnt,
+                               cncur->cncur_reverse ? cn_kv_cmp_rev : cn_kv_cmp, &lcur->cnlc_bh);
+        if (ev(err))
+            return err;
+    }
+
+    return err;
+}
+
+merr_t
+cn_tree_cursor_create(struct cn_cursor *cur)
+{
+    int i;
+    merr_t err = 0;
+    void *lock;
+    struct cn_level_cursor *lcur;
+    struct cn_tree *tree = cn_get_tree(cur->cncur_cn);
+
+    for (i = 0; i < NUM_LEVELS; i++) {
+        size_t kvref_tab_cnt = 1024 / sizeof(struct kvref);
+
+        lcur = &cur->cncur_lcur[i];
+        lcur->cnlc_cncur = cur;
+        lcur->cnlc_level = i;
+        if (!lcur->cnlc_kvref_tab)
+            lcur->cnlc_kvref_tab = table_create(kvref_tab_cnt, sizeof(struct kvref), false);
+
+        lcur->cnlc_esrcc = 64;
+        lcur->cnlc_esrcv = malloc(lcur->cnlc_esrcc * sizeof(*lcur->cnlc_esrcv));
+
+        if (!lcur->cnlc_kvref_tab || !lcur->cnlc_esrcv) {
+            err = merr(ENOMEM);
+            goto out;
+        }
+    }
+
+    lcur = &cur->cncur_lcur[0];
+
+    rmlock_rlock(&tree->ct_lock, &lock);
+    err = cn_tree_kvset_refs(tree->ct_root, lcur);
+    rmlock_runlock(lock);
```

Comment:
1. Why get refs on all root node kvsets instead of just the ones that have keys in the cursor's range?
2. Why is the root node inspected here but not the leaf node?